training/deepspeech_training/train.py

```diff
@@ -75,145 +64,101 @@ def create_overlapping_windows(batch_x):
-                                                 direction='unidirectional',
-                                                 dtype=tf.float32)
-        rnn_impl_cudnn_rnn.cell = fw_cell
+        self.dropout = tf.keras.layers.Dropout(dropout_rate) if dropout_rate != 0. else tf.keras.layers.Lambda(lambda x: x)
+        self.layer_norm = tf.keras.layers.LayerNormalization() if layer_norm else tf.keras.layers.Lambda(lambda x: x)
 
-    output, output_state = rnn_impl_cudnn_rnn.cell(inputs=x,
-                                                   sequence_lengths=seq_length)
+    def call(self, x):
+        return self.dropout(self.layer_norm(super().call(x)))
 
-    return output, output_state
 
-rnn_impl_cudnn_rnn.cell = None
+class Model(tf.keras.Model):
+    def __init__(self, dropout_rates=None, overlap=True, unroll=False, masking=True):
+        super().__init__(name='DeepSpeech_v{}'.format(int(file_relative_read('GRAPH_VERSION').strip())))
 
+        if dropout_rates is None:
+            dropout_rates = defaultdict(lambda: 0.)
 
-def rnn_impl_static_rnn(x, seq_length, previous_state, reuse):
-    with tfv1.variable_scope('cudnn_lstm/rnn/multi_rnn_cell'):
-        # Forward direction cell:
-        fw_cell = tfv1.nn.rnn_cell.LSTMCell(Config.n_cell_dim,
-                                            forget_bias=0,
-                                            reuse=reuse,
-                                            name='cudnn_compatible_lstm_cell')
+        self.overlap_layer = CreateOverlappingWindows()
+        self.overlap_op = self.overlap_layer if overlap else self.overlap_layer.reshape
 
-        # Split rank N tensor into list of rank N-1 tensors
-        x = [x[l] for l in range(x.shape[0])]
+        self.mask = tf.keras.layers.Masking() if masking else tf.keras.layers.Lambda(lambda x: x)
 
-        output, output_state = tfv1.nn.static_rnn(cell=fw_cell,
-                                                  inputs=x,
-                                                  sequence_length=seq_length,
-                                                  initial_state=previous_state,
-                                                  dtype=tf.float32,
-                                                  scope='cell_0')
+        def clipped_relu(x):
+            return tf.minimum(tf.nn.relu(x), FLAGS.relu_clip)
 
-        output = tf.concat(output, 0)
+        self.dense1 = Dense(Config.n_hidden_1, dropout_rate=dropout_rates[0], activation=clipped_relu, layer_norm=FLAGS.layer_norm, name='dense1')
+        self.dense2 = Dense(Config.n_hidden_2, dropout_rate=dropout_rates[1], activation=clipped_relu, layer_norm=FLAGS.layer_norm, name='dense2')
+        self.dense3 = Dense(Config.n_hidden_3, dropout_rate=dropout_rates[2], activation=clipped_relu, layer_norm=FLAGS.layer_norm, name='dense3')
 
-    return output, output_state
+        self.lstm = tf.keras.layers.LSTM(Config.n_cell_dim, return_sequences=True, return_state=True, unroll=unroll)
```

Comment:
no more several kind of implementations, how does this copes with some cudnn specific things? I can't recall exactly, but I remember there were some limitations in Keras's LSTM blocking us: can we use BlockFused as we used to?

Reply:
The Keras layer is supposed to automatically choose an applicable fast implementation, such as CuDNN RNN.

Reply:
Oh, I can see how this can fail in many ways. :)